rustfmt breaks code by removing `unsafe`

### STR

``` rust
fn main() {}

pub struct Peripherals {}

static mut PERIPHERALS: bool = false;

impl Peripherals {
    pub fn all() {
        foo(|| unsafe {
            if PERIPHERALS {
                loop {}
            } else {
                PERIPHERALS = true;
            }
        })
    }
}

fn foo<F>(_f: F)
where
    F: FnOnce(),
{
}
```

``` console
$ cargo check && echo OK
OK

$ rustfmt src/main.rs

$ cargo check
error[E0133]: use of mutable static requires unsafe function or block
(..)

$ cat src/main.rs
```

``` rust
fn main() {}

pub struct Peripherals {}

static mut PERIPHERALS: bool = false;

impl Peripherals {
    pub fn all() {
        foo(|| {
            if PERIPHERALS {
                loop {}
            } else {
                PERIPHERALS = true;
            }
        })
    }
}

fn foo<F>(_f: F)
where
    F: FnOnce(),
{
}
```

### Meta

```
$ rustfmt -V
0.2.15-nightly (f987946 2017-11-20)
```

--- a/src/closures.rs
+++ b/src/closures.rs
@@ -256,7 +256,9 @@ pub fn rewrite_last_closure(
 ) -> Option<String> {
     if let ast::ExprKind::Closure(capture, ref fn_decl, ref body, _) = expr.node {
         let body = match body.node {
-            ast::ExprKind::Block(ref block) if is_simple_block(block, context.codemap) => {
+            ast::ExprKind::Block(ref block)
+                if !is_unsafe_block(block) && is_simple_block(block, context.codemap) =>
+            {
                 stmt_expr(&block.stmts[0]).unwrap_or(body)
             }
             _ => body,

--- a/tests/source/closure.rs
+++ b/tests/source/closure.rs
@@ -188,3 +188,13 @@ fn issue1524() {
     let f = |x| {x};
     let f = |x| x;
 }
+
+fn issue2171() {
+    foo(|| unsafe {
+        if PERIPHERALS {
+            loop {}
+        } else {
+            PERIPHERALS = true;
+        }
+    })
+}

--- a/tests/target/closure.rs
+++ b/tests/target/closure.rs
@@ -220,3 +220,13 @@ fn issue1524() {
     let f = |x| x;
     let f = |x| x;
 }
+
+fn issue2171() {
+    foo(|| unsafe {
+        if PERIPHERALS {
+            loop {}
+        } else {
+            PERIPHERALS = true;
+        }
+    })
+}

